### ISSUE TITLE
Incorrect trafiic counters in CT map

### ISSUE THREAD

**Issue by MadMax240484** [cilium/cilium#36311]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.15.11 and lower than v1.16.0

### What happened?

Hi. I have cluster with cilium 1.16 and want to get counters from flow in CT maps.
In cilium version below 1.16 flow has separated rx/tx counter in flow record, all works fine.
After [commit](https://github.com/cilium/cilium/commit/641f1f892f885d586ef4e7252483f5e29b2caeb4#diff-dcc9f01af9e3e84d0d2ec148b2bd1691debe9774a62449cf6eab13398f0d7a90R78) direction counters was depracated. 
As i understand each direction record in CT map now have to collect only own traffic.
But the problem is that both record (IN/OUT) for the same flow update the same data.(total data from flow)
![Image](https://github.com/user-attachments/assets/2e5399d9-d83a-416d-b09c-cb4aecc55600)
I attach the example of easy nginx pod with LB and it's flow record updates

### How can we reproduce the issue?

1. Deploy any service/pod
2. In cilium 1.16 enable --bpf-conntrack-accounting-enabled flag via cli in cilium pod **cilium config ConntrackAccounting=true**
3. Start some TCP session
4. List CT map filtered by IP

### Cilium Version

1.16.1

### Kernel Version

5.14.0-427.16.1.el9_4.x86_64

### Kubernetes Version

v1.30.4

### Regression

1.15

### Relevant log output

```shell

```

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

**Comment by joestringer** [comment]
This seems like a consequence of the changes introduced in https://github.com/cilium/cilium/pull/31087, but I'm not sure it's a bug - it seems like intentional behaviour. Perhaps it could have been better announced but as far as I read from the referenced commit there was no intent to retain the ability to distinguish sent/received traffic counters. cc @borkmann

**Comment by MadMax240484** [comment]
> This seems like a consequence of the changes introduced in [#31087](https://github.com/cilium/cilium/pull/31087), but I'm not sure it's a bug - it seems like intentional behaviour. Perhaps it could have been better announced but as far as I read from the referenced commit there was no intent to retain the ability to distinguish sent/received traffic counters. cc [@borkmann](https://github.com/borkmann)

Thanks for info, Joe. But i think that if this behaviour is intentional, very difficult to understand how many packets was in each direction. This information is very helpful for troubleshooting.

**Comment by joestringer** [comment]
I agree, I can see how that's useful for troubleshooting. I'll assign this to @cilium/sig-datapath for further discussion.

**Comment by MadMax240484** [comment]
Any updates?

**Comment by MadMax240484** [comment]
not stale, waiting for fix

**Comment by borkmann** [comment]
> > waiting for fix
> 
> It's not a bug, but an intentional change: the [commit you pointed to](https://github.com/cilium/cilium/pull/31087/commits/30bf7d5d87167ec5ce3aac7a2f2907621fb85fa6) intentionally squashes rx+tx together, though I do understand the frustration about removing existing features. Please note, however, that even before that change, the separate counters didn't work as users might expect: [rx_bytes weren't accounted separately for CT_SERVICE](https://github.com/cilium/cilium/pull/31087/commits/30bf7d5d87167ec5ce3aac7a2f2907621fb85fa6#diff-78122dbc1bf6a53a08f02c035e071f869eaa123b52822d789b79c04ff1cbdf69L936-L938), which was more confusing.
> 
> [@borkmann](https://github.com/borkmann), I see that your commit freed up space in struct ct_entry. Could you remind me what was the plan? Was the new free space repurposed for something else?
> 
> Is it possible to do __sync_fetch_and_add on a 32-bit scalar in BPF? As accounting is turned off by default, people turn it on for debug purposes, so maybe 32-bit counters would be enough for that, if that works at all...

iirc on old kernel this wouldn't work, on new ones potentially yes wrt atomics. The CT counters were consolidated to free up space for future use in the CT entry struct, and accounting got turned off by default few releases ago to remove the cost of the atomics in fast path. The CT map never had a stable API guarantee; and besides what Max said earlier wrt multiple entries and directions it doesn't make sense to have rx/tx accounting there, but rather consolidated counters.

**Comment by MadMax240484** [comment]
> Please note, however, that even before that change, the separate counters didn't work as users might expect: [rx_bytes weren't accounted separately for CT_SERVICE](https://github.com/cilium/cilium/pull/31087/commits/30bf7d5d87167ec5ce3aac7a2f2907621fb85fa6#diff-78122dbc1bf6a53a08f02c035e071f869eaa123b52822d789b79c04ff1cbdf69L936-L938), which was more confusing
Of course i understand that CT map has different types of record. But not all of them are useful for throublshooting, especcialy if you have to know how many packets was forwarded from client to pod and vice versa. You only need record with "Protocol +IN/OUT " flags for session.


> The CT map never had a stable API guarantee; and besides what Max said earlier wrt multiple entries and directions it doesn't make sense to have rx/tx accounting there, but rather consolidated counters.

Daniel, i understand this. Sorry, if i'm wrong, but it is difficult to guess what for only total amount of packets/bytes will be useful collect?

**Comment by MadMax240484** [comment]
> > Please note, however, that even before that change, the separate counters didn't work as users might expect: [rx_bytes weren't accounted separately for CT_SERVICE](https://github.com/cilium/cilium/pull/31087/commits/30bf7d5d87167ec5ce3aac7a2f2907621fb85fa6#diff-78122dbc1bf6a53a08f02c035e071f869eaa123b52822d789b79c04ff1cbdf69L936-L938), which was more confusing
> 
> Of course i understand that CT map has different types of records. But not all of them are useful for throublshooting, especcialy if you have to know how many packets was forwarded from client to pod and vice versa. You only need record with "Protocol +IN/OUT " flags for session.
> 
> > The CT map never had a stable API guarantee; and besides what Max said earlier wrt multiple entries and directions it doesn't make sense to have rx/tx accounting there, but rather consolidated counters.
> 
> Daniel, i understand this. Sorry, if i'm wrong, but it is difficult to guess what for only total amount of packets/bytes will be useful collect? Only  for simple statustics, not more. Previous case allow to collect data to Netflow(NSEL), for example.

**Comment by MadMax240484** [comment]
> > > Please note, however, that even before that change, the separate counters didn't work as users might expect: [rx_bytes weren't accounted separately for CT_SERVICE](https://github.com/cilium/cilium/pull/31087/commits/30bf7d5d87167ec5ce3aac7a2f2907621fb85fa6#diff-78122dbc1bf6a53a08f02c035e071f869eaa123b52822d789b79c04ff1cbdf69L936-L938), which was more confusing
> > 
> > 
> > Of course i understand that CT map has different types of records. But not all of them are useful for throublshooting, especcialy if you have to know how many packets was forwarded from client to pod and vice versa. You only need record with "Protocol +IN/OUT " flags for session.
> > > The CT map never had a stable API guarantee; and besides what Max said earlier wrt multiple entries and directions it doesn't make sense to have rx/tx accounting there, but rather consolidated counters.
> > 
> > 
> > Daniel, i understand this. Sorry, if i'm wrong, but it is difficult to guess what for only total amount of packets/bytes will be useful collect? Only  for simple statustics, not more. Previous configuration allow to collect data to Netflow(NSEL), for example.

**Comment by MadMax240484** [comment]
> Please note, however, that even before that change, the separate counters didn't work as users might expect: [rx_bytes weren't accounted separately for CT_SERVICE](https://github.com/cilium/cilium/pull/31087/commits/30bf7d5d87167ec5ce3aac7a2f2907621fb85fa6#diff-78122dbc1bf6a53a08f02c035e071f869eaa123b52822d789b79c04ff1cbdf69L936-L938), which was more confusing

 Of course i understand that CT map has different types of records. But not all of them are useful for throublshooting, especcialy if you have to know how many packets was forwarded from client to pod and vice versa. You only need record with "Protocol +IN/OUT " flags for session.

 > The CT map never had a stable API guarantee; and besides what Max said earlier wrt multiple entries and directions it doesn't make sense to have rx/tx accounting there, but rather consolidated counters.

Daniel, i understand this. Sorry, if i'm wrong, but it is difficult to guess what for only total amount of packets/bytes will be useful collect? Only  for simple statustics, not more. Previous configuration allow to collect data to Netflow(NSEL), for example.
